src/ol/proj.js

```diff
@@ -625,6 +626,66 @@ export function fromUserExtent(extent, destProjection) {
   return transformExtent(extent, userProjection, destProjection);
 }
 
+/**
+ * Creates a safe coordinate transform function from a coordinate transform function.
+ * "Safe" means that it can handle wrapping of x-coordinates for global projections,
+ * and that coordinates exceeding the projection validity extent's range will be
+ * clamped to the validity range.
+ * @param {Projection} sourceProj Source projection.
+ * @param {Projection} destProj Destination projection.
+ * @param {function(import("./coordinate.js").Coordinate): import("./coordinate.js").Coordinate} forward Transform function (source to destiation).
+ * @param {function(import("./coordinate.js").Coordinate): import("./coordinate.js").Coordinate} inverse Transform function (destiation to source).
+ * @return {function(import("./coordinate.js").Coordinate): import("./coordinate.js").Coordinate} Safe transform function (source to destiation).
+ */
+export function createSafeCoordinateTransform(
+  sourceProj,
+  destProj,
+  forward,
+  inverse
+) {
+  return function (coord) {
+    let x, y, worldsAway;
+    if (sourceProj.canWrapX()) {
+      const sourceExtent = sourceProj.getExtent();
+      const sourceExtentWidth = getWidth(sourceExtent);
+      worldsAway = getWorldsAway(coord, sourceProj, sourceExtentWidth);
+      if (worldsAway && sourceExtent) {
+        // Move x to the real world
+        x = coord[0] - worldsAway * sourceExtentWidth;
+      }
+    }
+    let transformed = forward(x === undefined ? coord : [x, coord[1]]);
+    const destExtent = destProj.getExtent();
+    if (!isFinite(transformed[0]) || !isFinite(transformed[1])) {
+      // Try to recover from out-of-bounds transform
+      if (destExtent) {
+        const corner1 = inverse(destExtent.slice(0, 2));
+        const corner2 = inverse(destExtent.slice(2, 4));
+        const x1 = corner1[0];
+        const x2 = corner2[0];
+        const y1 = corner1[1];
+        const y2 = corner2[2];
+        if (isFinite(x1) && isFinite(x2)) {
+          x = clamp(
+            x == undefined ? coord[1] : x,
+            Math.min(x1, x2),
+            Math.max(x1, x2)
+          );
+        }
+        if (isFinite(y1) && isFinite(y2)) {
+          y = clamp(coord[1], Math.min(y1, y2), Math.max(y1, y2));
```

Comment:
I think you'd need to transform all four corners to determine which resulted in the min/max y.  As this is written, if the top-left corner transforms to a higher y-value than the top-right corner, the max y will be smaller than it should.